src/ucp/core/ucp_am.c

```diff
@@ -1199,11 +1200,11 @@ ucp_am_handler_common(ucp_worker_h worker, ucp_am_hdr_t *am_hdr, size_t hdr_size
         return UCS_OK;
     }
 
-    status = ucp_recv_desc_init(worker, data,
-                                total_length - hdr_size - am_hdr->header_length,
-                                0,
-                                UCT_CB_PARAM_FLAG_DESC, /* pass as a const */
-                                0, 0, -hdr_size, &desc);
+    payload_offset = hdr_size + am_hdr->header_length;
+    status         = ucp_recv_desc_init(worker, data,
+                                        total_length - payload_offset, 0,
+                                        UCT_CB_PARAM_FLAG_DESC, /* pass as a const */
+                                        0, 0, -payload_offset, &desc);
```

Comment:
it means that before this PR, user got pointer to user_hdr instead of pointer to payload?

Reply:
no, why?

Reply:
@yosefe, added picture to the description of the issue

Reply:
so the problem was passing an incorrect pointer to ucs_free() ?

Reply:
yes

Reply:
> yes

should be described like this:
### Why 
Fix invalid pointer free when releasing AM descriptor with non-empty user header, since ucp_am_data_release() does not consider user header size when calculating descriptor pointer from data handle

### How
Save the distance from "data" pointer to malloc'd buffer pointer in ~~payload_offset~~ am_malloc_offset field

Reply:
updated